### PR TITLE
implementation for using structured output to filter out invalid answers in refine synthesis

### PR DESCRIPTION
**Pull request by sourabhdesai**
# Description

implementation for using structured output to filter out invalid answers in refine synthesis.

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: llama_index/program/llm_program.py [hunk 1]
```diff
@@ -64,11 +70,11 @@ def __call__(
         **kwargs: Any,
     ) -> BaseModel:
         prompt_with_parse_instrs_tmpl = self._output_parser.format(
-            self._prompt.original_template
+            self._prompt.format(**kwargs)
```
Comment (Disiok) [inline llama_index/program/llm_program.py:73]: curious: why this change?

File: llama_index/response_synthesizers/refine.py [hunk 2]
```diff
@@ -25,11 +76,24 @@ def __init__(
         refine_template: Optional[RefinePrompt] = None,
         streaming: bool = False,
         verbose: bool = False,
+        structured_answer_filtering: bool = False,
+        program_factory: Optional[Callable[[Prompt], BasePydanticProgram]] = None,
```
Comment (Disiok) [inline llama_index/response_synthesizers/refine.py:80]: nit: it's not immediately obvious to me when and how to set `program_factory`

Comment (Disiok) [reply llama_index/response_synthesizers/refine.py:80]: Oh I see it's useful for testing.

**Review by Disiok — APPROVED** [review]
Super nice!

File: llama_index/response_synthesizers/refine.py (hunk 2 above)
Comment (sourabhdesai) [reply llama_index/response_synthesizers/refine.py:80]: Its def useful for testing but I can also imagine a real use case for it. It opens up the possibility for non-OpenAI function calling users to provide their own `Program` implementations using their own custom prompts.

**Comment by logan-markewich** [comment]
Awesome, this looks just about ready to merge! Two more notes

- looks like the changelog entry for this PR got a bit lost in the past 😉 
- Can we also link the notebook you added in `/home/loganm/llama_index_proper/llama_index/docs/core_modules/query_modules/response_synthesizers/modules.md`

**Review by logan-markewich — APPROVED** [review]
Yea this seems good to me! Awesome!

Tried a few examples locally with the PG essay -- seemed hard to actually find a query where structured_filtering returned a different response, but did find a few!

I think a great follow-up PR to this would be adding re-ranking filtering instead of structured responses -- I might look into this more 👍🏻
